UIPS/UIP-0133.md

```diff
@@ -0,0 +1,202 @@
+- Limited visibility into test coverage via CI
+
+By unifying standards and aligning expectations across the ecosystem, we address all of the above. This UIP captures and formalizes the best emerging practices from both contexts and lays a foundation for continued iteration.
+
+# Specification
+
+This document formalizes the following testing standards for all code in the `%base` desk.
+
+## Arvo Standards
+
+Arvo’s functional architecture makes it well-suited to precise unit testing and isolated module validation. Cores expose well-defined interfaces, allowing developers to reason about behavior and test correctness without relying on dynamic analysis or anomaly detection. This design emphasizes thorough unit and integration tests as the primary means of ensuring robustness.
+
+### Unit Testing
+
+Unit testing ensures the correct functionality of isolated code modules. Arvo’s structure encourages small, focused tests that validate each core’s behavior in isolation.
+
+#### Hoon Language & Libraries
+
+All Hoon language features and standard library functions—including `hoon.hoon`, `lull.hoon`, `zuse.hoon`, and `arvo.hoon`, as well as all files in `/lib`—MUST have unit tests.
+
+Tests should use the test thread and associated library, optionally supplemented with `%quiz`. Test files should reside in `/tests` and follow the naming convention:
+
+    urbit/tests/[base-desk-file-path].hoon
+
+For partial tests of individual arms, append the arm name to the path.
+
+**Example:**  
+This test verifies addition behavior, including identity and basic arithmetic. It should be located at `tests/sys/hoon/math/add.hoon`.
+
+```
+/+  *test
+|%
+::
+::  Test addition (+)
+::
+++  test-add
+  ;:  weld
+    ::  Checks standard addition
+    ::
+    %+  expect-eq
+      !>  2
+      !>  (add 1 1)
+    ::  Checks identity property (0 + n = n)
+    ::
+    %+  expect-eq
+      !>  5
+      !>  (add 0 5)
+  ==
+--
+```
+
+#### Agents
+
+Agent tests should use the `test-agent` library for structured verification of state and cards.
+
+**Example:**  
+This test builds the `%time` agent, initializes it, pokes it, and checks for the expected `%wait` card. It should be located at `tests/app/time.hoon`.
+
+```
+/+  *test-agent
+/=  time-agent  /app/time
+|%
+++  test-poke
+  %-  eval-mare
+  =/  m  (mare ,~)
+  ^-  form:m
+  ;<  *  bind:m  (do-init %time time-agent)
+  ;<  caz=(list card)  bind:m  (do-poke noun+!>(~))
+  ;<  =bowl  bind:m  get-bowl
+  %+  ex-cards
+    caz
+  :~  (ex-arvo /(scot %da now.bowl) %b %wait `@da`+(now.bowl))
+  ==
+--
+```
+
+#### Vanes
+
+Each vane should have unit tests covering its API. The style used in [current Eyre tests](https://github.com/urbit/urbit/blob/develop/tests/sys/vane/eyre.hoon) is preferred and should be adopted across vanes when possible. A universal `test-vane` library may be impractical, but per-vane test libraries are encouraged to maintain consistency and reduce duplication.
```

Comment:
Should probably prefer permalinks (iow links to files at specific commits) for durability.